### ISSUE TITLE
Performance hit when using callback

### ISSUE THREAD

**Issue by fredcallaway** [JuliaNLSolvers/Optim.jl#907]
When passing a callback function to a univariate optimization, I find that a little over half of the runtime is in the call to `trace!`, specifically the use of Dict. This is unfortunate given that my early stopping criterion only depends on the best-found value (a common use case, I think).

I think a good general-purpose solution here would be to provide a `target_value` argument such that the optimization stops as soon as a value lower than the target is found (without invoking a callback function). I will probably implement this for my own use, but if there's interest I can make a pull request.

**Comment by fredcallaway** [comment]
Implementation is super easy. Most of the work is updating interface functions. Also it currently requires float arguments (because I wasn't sure how to cast to float when the argument can be Nothing.

https://github.com/fredcallaway/Optim.jl/commit/a66a9b29e46d42342fd554d0564b0df1c6bcf73e

**Comment by pkofod** [comment]
Yes the target value has certainly been mentioned before. You don't have to annotate it at all I think.

**Comment by fredcallaway** [comment]
The issue is that we need to cast an Int argument to Float64 (but a BigFloat should stay a BigFloat). This is currently done with `Tf = float(T)`, but float(Nothing) doesn't work. Of course this is not so hard to get around, I was just lazy in the first pass.

Would you be interested in a pull request?

**Comment by pkofod** [comment]
I don't think it matters since it's just a comparison. You can compare `0.4` to `4` just fine.

**Comment by pkofod** [comment]
In the successor code to what's currently in Optim.jl I have this and it's called. I haven't ported the univariate solvers yet though. https://github.com/pkofod/NLSolvers.jl/blob/65034bb76cbf91e22784bd0156f28b85416adb2f/src/optimize/problem_types.jl#L126
